### DISCUSSION TITLE
Parent Model doesn't raise Validation Error for a different field type

### DISCUSSION THREAD

**Discussion by OlegZv** [fastapi/fastapi#10483] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class ModelA(BaseModel):
    field_a: str
    field_b: str
    field_c: list[str]


class ModelB(ModelA):
    field_c: dict[str, str] = {}


@app.get('/test', response_model=ModelA)
def get_status():
    m = ModelB.model_validate({"field_a": "1", "field_b": "2", "field_c": {"3": "12"}})
    return m
```

### Description

I expect this to raise a validation error since the `ModelA.field_c` is a `list[str]` and the `m` instance has `field_c` as a `dict`.
Instead, the response comes out as Model B (with `field_c` as a `dict`) and produces this in the logs.

Test by calling `curl -X 'GET'  'http://localhost:8000/test' -H 'accept: application/json'`
```
.venv/lib/python3.10/site-packages/pydantic/type_adapter.py:356: UserWarning: Pydantic serializer warnings:
  Expected `list[str]` but got `dict` - serialized value may not be as expected
  return self.serializer.to_python(
```
The expected behavior is raising ResponseValidationError:
```
fastapi.exceptions.ResponseValidationError: 1 validation error:
  {'type': 'list_type', 'loc': ('response', 'field_c'), 'msg': 'Input should be a valid list', 'input': {'3': '12'}, 'url': 'https://errors.pydantic.dev/2.4/v/list_type'}
```

### Operating System

Linux

### Operating System Details

Distributor ID: Ubuntu
Description: Ubuntu 22.04.3 LTS
Release: 22.04
Codename: jammy

### FastAPI Version

0.103.0

### Pydantic Version

2.4.2

### Python Version

3.10.12

**Comment by pythonweb2** [comment]
Under the hood it appears that FastAPI is using pydantic type adapters to serialize the data. There seems to be something weird going on (maybe has to do with the inheritance in this situation). Below is an example of where I can reproduce this issue.

```python
from pydantic import BaseModel
from pydantic import TypeAdapter


class ModelA(BaseModel):
    field_a: str
    field_b: str
    field_c: list[str]


class ModelB(ModelA):
    field_c: dict[str, str] = {}


if __name__ == '__main__':
    t = TypeAdapter(ModelA)

    m = ModelB(
        field_a="1",
        field_b="2",
        field_c={"3": "12"}
    )
    print(m)

    x = t.validate_python(m, from_attributes=True)  # Returns and instance of model B?
    print(x)

    y = t.validate_python({"field_a": "foo", "field_b": "bar", "field_c": ["1", "2", "3"]})  # Works fine
    print(y)

    z = ModelA.model_validate(m, from_attributes=True)  # Returns and instance of model B?
    print(z)
```

However, without inheritance it seems to work fine, and I get this:

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for ModelA
field_c
  Input should be a valid list [type=list_type, input_value={'3': '12'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/list_type
```

It seems like a bug to me, but you would need to open the issue on the pydantic page. There may be some reasoning for this that I am not understanding.

**Comment by YuriiMotov** [comment, marked as answer]
Looks like Pydantic's issue:

```python
from pydantic import BaseModel, ValidationError
import pytest

class ModelA(BaseModel):
    field_a: str
    field_b: str
    field_c: list[str]

class ModelB(ModelA):
    field_c: dict[str, str] = {}

class ModelC(BaseModel):
    field_a: str
    field_b: str
    field_c: dict[str, str] = {}

INPUT = {"field_a": "1", "field_b": "2", "field_c": {"3": "12"}}

def test_validate_dict():
    with pytest.raises(ValidationError):
        ModelA.model_validate(INPUT)

def test_validate_object():
    b = ModelB.model_validate(INPUT)
    with pytest.raises(  # Failed: DID NOT RAISE <class 'pydantic_core._pydantic_core.ValidationError'>
        ValidationError
    ):
        ModelA.model_validate(b)

def test_validate_object_not_inherited():
    c = ModelC.model_validate(INPUT)
    with pytest.raises(ValidationError):
        ModelA.model_validate(c)
```

**Reply by OlegZv** [reply]
Thank you! I opened a bug with `pydantic` https://github.com/pydantic/pydantic/issues/11936
